pageserver/src/repository.rs

```diff
@@ -9,10 +9,29 @@ use std::time::Duration;
 use zenith_utils::lsn::{Lsn, RecordLsn};
 use zenith_utils::zid::ZTimelineId;
 
+// NOTE: Now this enum provides only a few basic states needed
+// to handle graceful shutdown.
+// In the future in can be expanded to treat other states.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum RepositoryState {
+    // The repository exists, but not yet initialized/loaded.
+    New,
+    // The repository is initialized and ready to serve queries.
+    Running,
+    // This state is set when graceful shutdown of the tenant/pageserver is requested.
+    // Repository working threads periodically check the state
+    // and stop their job when shutdown is requested.
+    ShutdownRequested,
+}
+
 ///
 /// A repository corresponds to one .zenith directory. One repository holds multiple
 /// timelines, forked off from the same initial call to 'initdb'.
 pub trait Repository: Send + Sync {
+    fn set_state(&self, state: RepositoryState);
```

Comment:
I've got an impression, that `state` could be kept outside of the repository's public api, similar to the proposal in https://github.com/zenithdb/zenith/pull/689#issuecomment-931216193

I see a few reasons for that:

* feels a bit odd to plug in the "download layer on demand" case into the picture: when we trigger a download, there's no timeline created yet and no files to create the timeline from, so it's hard to set any "Downloading" status to nothing :slightly_smiling_face: 
* seems like all `set_state` calls are located in `tenant_mgr`, where we have to lock to get a repository anyway
* all `get_state`/`wait_state_update` calls are done in `impl LayeredRepository` and all check for the shutdown flag only

To put it all together, feels like `tenant_mgr` could hold the state indeed in the repo map under the mutex and update it when needed?
The repo can get an `AtomitBoolean` flag for shutdown and leave the state management and waiting for the managing code